### ISSUE TITLE
Request to search for record changes in change history

### ISSUE THREAD

**Issue by kontrolldkaos** [vinyldns/vinyldns#960]
**Is your feature request related to a problem? Please describe.**
In some environments there are hundreds to thousands of changes per day, looking for past changes can be challenging.  The change history page is just one page with no search options. 

**Describe the solution you'd like**
Adding search options by fields: ID, User, record name, date range, and if status is completed or failed. 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by pauljamescleary** [comment]
Related to this is that we need an approach to archiving older changes.  @kontrolldkaos is all history imperative?  Should retention be a config option?

**Comment by remerle** [comment]
@pauljamescleary it definitely needs to be a config setting if there is support for archiving. Feedback that we've gotten is that all "significant" changes should be kept indefinitely. 

Significant changes were basically identified as anything but TTL changes. Everything else was needed for later analysis.

For enterprise users, they can come up with custom archiving strategies at the DB level. Ex: archive or replicate change records to a data lake for reporting and analysis.

As it stands the larger challenge is the unsearchable binary blobs, IMO.

**Comment by pauljamescleary** [comment]
@remerle makes sense, we can stand up new indexes on record changes and pull fields out of the blob.  We have done this previously via a migration script.  Ew, by the looks of it we have nothing broken out for mysql.  Not a fun migration for existing installs if there is a lot of data, but the code itself should be minimal.

We could add a `recordSetId` column+index, that would allow us searches by record type, record name, fqdn, and even group by joining through to `recordset` table.

In addition
- add a `userId` column+index 
- status column+index

Note: there appears to already be a `created` index on that table, so we should be fine to search by time already.

**Comment by kontrolldkaos** [comment]
> Related to this is that we need an approach to archiving older changes. @kontrolldkaos is all history imperative? Should retention be a config option?

Yes retention should be a config option. Some organizations may not need that long of retention, while others may. To archive, I'd look into stripping certain changes that aren't that important, mainly TTL changes/adjustments.

**Comment by kontrolldkaos** [comment]
and I should have read Ryan's replies...yep..he has it covered.

**Comment by pauljamescleary** [comment]
Ok, I think this is good to be implemented?  Perhaps we have a "Ready" label @remerle ?

1. Flyway migration to add the columns
2. A "programmatic" flyway migration that populates the columns (this will take a long time to run.  If done as a flyway migration, it will delay the restart of VinylDNS until it is complete which maybe ok for us due to our release pipeline.  Release notes should make this very clear
3. Add additional query support in the record change repository to take in the following fields
    1. user id
    2. record mask - to be used in a `LIKE` clause; our convention is to support wildcards and replace with `%`
    3. start time (Instant)
    4. end time (Instant)
    5. status

@kontrolldkaos what do you mean by "ID", are you talking about batch changes / DNS requests?

**Comment by remerle** [comment]
@pauljamescleary I think we might want a migration script as part of our migration path.  I'd be concerned for organizations that have 10s of millions of recordset changes waiting on Flyway to do it's thing and holding up the startup of the app.  Having the ability to non-destructively migrate production data outside of the release pipeline might be a good option.

I think this is conceptually pretty straightforward and ready to go.

**Comment by kontrolldkaos** [comment]
> @kontrolldkaos what do you mean by "ID", are you talking about batch changes / DNS requests?

Yes I am talking about batch change ID.

**Comment by pauljamescleary** [comment]
@remerle I haven't looked into it, but perhaps we can have flyway run certain migrations out of band, as opposed to on startup by default.  Having that lever to override will allow people to proceed carefully.  Need to fully document that in the Release Notes

**Comment by remerle** [comment]
Removing this from `ready` in favor of a wholly revisiting the BLOBs and pulling data up into columns (#975). Implementing #975 will open the door to many features such as this.

**Comment by remerle** [comment]
Closing in favor of #975
